src/circuit/gadgets/constraints.rs

```diff
@@ -148,6 +186,29 @@ pub fn sub<F: PrimeField, CS: ConstraintSystem<F>>(
     Ok(res)
 }
 
+/// Adds a constraint to CS, enforcing a product relationship between the allocated numbers a, b, and product.
+///
+/// a * b = product - c
+pub fn linear<F: PrimeField, A, AR, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    annotation: A,
+    a: &AllocatedNum<F>,
+    b: &AllocatedNum<F>,
+    c: &AllocatedNum<F>,
+    l: &AllocatedNum<F>,
```

Comment:
I assume `l` is what the comments calls `product` everywhere. Let's change one or the other so they match throughout. Is product really the right name?

The documentation above is also a little confusing. A fuller explanation might help explain why this function is called `linear`.

Reply:
Renamed and updated documentation.